pkgs/development/compilers/tinygo/default.nix

```
@@ -1,64 +1,74 @@
-{ lib, buildGoModule, fetchFromGitHub, llvm, clang-unwrapped, lld, avrgcc
-, avrdude, openocd, gcc-arm-embedded, makeWrapper, fetchurl }:
+{ lib
+, buildGoModule
+, fetchFromGitHub
+, makeWrapper
+, llvm
+, clang-unwrapped
+, lld
+, avrgcc
+, go
+, gcc-arm-embedded
+, avrdude
+, openocd
+}:
 
-let main = ./main.go;
-    gomod = ./go.mod;
-in
-buildGoModule rec {
+let
+  llvmMajor = lib.versions.major llvm.version;
+in buildGoModule rec {
   pname = "tinygo";
-  version = "0.16.0";
+  version = "0.21.0";
 
   src = fetchFromGitHub {
     owner = "tinygo-org";
-    repo = "tinygo";
+    repo = pname;
     rev = "v${version}";
-    sha256 = "063aszbsnr0myq56kms1slmrfs7m4nmg0zgh2p66lxdsifrfly7j";
+    sha256 = "sha256-ZAUvOKApwOIKWINta9jEs/M32pHgW86+a0Y1jHEMupk=";
     fetchSubmodules = true;
   };
 
-  overrideModAttrs = (_: {
-      patches = [];
-      preBuild = ''
-      rm -rf *
-      cp ${main} main.go
-      cp ${gomod} go.mod
-      chmod +w go.mod
-      '';
-  });
+  vendorSha256 = "sha256-fD9/hPRXIT7cPljA+QNWDRESqHXeCbtWHMq12280R1g=";
 
-  preBuild = "cp ${gomod} go.mod";
-
-  postBuild = "make gen-device";
+  nativeBuildInputs = [ makeWrapper ];
+  buildInputs = [ llvm clang-unwrapped ];
 
-  vendorSha256 = "12k2gin0v7aqz5543m12yhifc0xsz26qyqra5l4c68xizvzcvkxb";
+  subPackages = [ "." ];
 
   doCheck = false;
 
-  prePatch = ''
+  # We need to set GOROOT to import and compile
+  # standard go libraries with tinygo
+  allowGoReference = true;
+
+  postPatch = ''
+    patchShebangs lib/wasi-libc
     sed -i s/', "-nostdlibinc"'// builder/builtins.go
     sed -i s/'"-nostdlibinc", '// compileopts/config.go builder/picolibc.go
   '';
 
-  subPackages = [ "." ];
-  nativeBuildInputs = [ makeWrapper ];
-  buildInputs = [ llvm clang-unwrapped ];
-  propagatedBuildInputs = [ lld avrgcc avrdude openocd gcc-arm-embedded ];
+  postBuild = ''
+    make gen-device
+  '';
 
   postInstall = ''
-    mkdir -p $out/share/tinygo
+    mkdir -p $out/share/tinygo $out/libexec/tinygo
     cp -a lib src targets $out/share/tinygo
```

You don't need to copy that much! Most of the files in lib are not necessary. You can check the `build/releases` target in the Makefile to see which files are. In fact, you could perhaps run `make build/release` to copy exactly these files to build/release and then copy lib out of it: this list of files tends to change over time.

```suggestion
    make build/release
    cp -a build/release/tinygo/lib src targets $out/share/tinygo
```